### PR TITLE
Unify encoding issues

### PR DESCRIPTION
**Pull request by mitch000001**
This introduces middlewares for the transport package. A used transport can now 
be composed of multiple ones. E.g.:

Before there was only one HttpsTransport which performed the transport via HTTPS 
but also encoded the request/decoded the response with Base64 encoding.
Now there is only a HttpsTransport which performs just the HTTPS transport and 
a separate Base64 Middleware which applies the encoding on top of the wrapped 
httpsTransport.

Also introduced was a UTF8Encoding middleware, but due to encrypted messages and 
binary data withing the payloads it seems quite impossible to apply it.

The former combined `transport.HttpsTransport` was renamed to 
`transport.HttpsBase64Transport`. 
Its returned response was also altered not to use `transport.ReadResponse` but to 
just return the response as is.

`dialog` now performs the task of reading the response so it can be used to 
extract the segments within it.

On the way of introducing middleware there was a transition away from holding
the raw bytes within `transport.Request` and `transport.Response` to a much
cleaner and more composable approach to use an `io.ReadCloser` as `Body`.

`transport.Response` still has the raw response in the field `MarshaledResponse`
but users are strongly disencouraged to use it as it will go away when the
internal users are transitioned to use the `Body` field instead.

One change also affects the encoding from and to ISO8859_1. Before there was a
manual code fragment converting the bytes to and from. This was replaced by
the encoding found in `golang.org/x/text/encoding` to rely on a much more used
and tested encoding. If there are any issued with the `convert` packages, or
if the code from that package panics for some reasons, feel free to open an 
issue.

I apologize for these breaking changes, but all higher level caller 
(package client, dialog) should not be affected in any way. 
All users of the `transport` package are strongly encouraged to use the new
Transports / Middlewares instead of the old combined Transport.

### CODE REVIEW

File: charset/convert.go [hunk 1]
```diff
@@ -3,23 +3,27 @@ package charset
...
+	return b
+}
+
+func ToUtf8(iso8859_1_buf []byte) string {
+	decoder := charmap.ISO8859_1.NewDecoder()
+	// TODO: propagate errors
+	return string(mustConvert(decoder.Bytes(iso8859_1_buf)))
 }
 
 func ToISO8859_1(utf8String string) []byte {
```
Comment (nicolai86) [inline charset/convert.go]: why the suffix `_1`? is there a meaning to this? Why does the exported function not have a doc?

File: client/testmessage.go [hunk 2]
```diff
@@ -6,18 +6,27 @@ import (
...
+		panic(err)
+	}
+	return []byte(str)
+}
+
 func encryptedTestMessage(dialogID string, encryptedData ...string) []byte {
+	iso8859Encoder := charmap.ISO8859_1.NewEncoder()
 	encryptionHeader := "HNVSK:998:2:+998+1+1::0+1:20150713:173634+2:2:13:@8@\x00\x00\x00\x00\x00\x00\x00\x00:5:1:+280:10000000:12345:V:0:0+0+'"
-	encryptionData := fmt.Sprintf("HNVSD:999:1:+@%d@%s'", len(charset.ToISO8859_1(strings.Join(encryptedData, ""))), charset.ToISO8859_1(strings.Join(encryptedData, "")))
+	encryptionData := fmt.Sprintf("HNVSD:999:1:+@%d@%s'", len(charset.ToISO8859_1(strings.Join(encryptedData, ""))), mustEncode(iso8859Encoder.String(strings.Join(encryptedData, ""))))
```
Comment (nicolai86) [inline client/testmessage.go]: I think pulling this line apart with multiple temporary variables would increase readability -> FTW

File: transport/https.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	var reader io.Reader
+	if httpResponse.StatusCode == http.StatusOK {
+		reader = base64.NewDecoder(base64.StdEncoding, httpResponse.Body)
+	} else {
+		reader = httpResponse.Body
+	}
+	return &Response{Body: ioutil.NopCloser(reader), Request: request}, nil
+}
+
+func NewHttpsTransport() Transport {
```
Comment (nicolai86) [inline transport/https.go]: two cents in terms of naming:

1. the package is called `transport`, so maybe just calling this `New` would be more appropriate, and moving the entire package into `transport/https`.
2. if the name contains `https`, it should be `NewHTTPSTransport`, as it's an abbreviation. 

also, comment are rare these days :)

File: charset/convert.go (hunk 1 above)
Comment (mitch000001) [inline charset/convert.go]: The reason for `_1` is that there are multiple encodings defined in ISO 8859. `_1` is commonly know as Latin1.

File: transport/https.go (hunk 3 above)
Comment (mitch000001) [inline transport/https.go]: 👍

File: charset/convert.go [hunk 4]
```diff
@@ -3,23 +3,41 @@ package charset
...
+}
+
+// ToUtf8 decodes the provided buffer from the ISO8859_1 encoding to UTF8.
+// The result is returned as string.
+// Any error from the decoder will cause a panic.
+//
+// This function hasn't change its signature as it is used in almost any package.
+// To address the issues with panicing errors this package will most likely get
+// a new decoding function which return possible decoding errors.
+func ToUtf8(iso8859_1_buf []byte) string {
```
Comment (nicolai86) [inline charset/convert.go]: Acronyms should be all upper case - see go naming conventions, e.g. https://talks.golang.org/2014/names.slide#6.

in this case, `ToUTF8`.

Comment (nicolai86) [inline charset/convert.go]: also, prefer mixedCase to snake_case, even for variable names.

File: transport/https/https.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+// encoding.
+func NewBase64() transport.Transport {
+	return &HttpsBase64Transport{
+		httpClient: http.DefaultClient,
+	}
+}
+
+// A HttpsTransport inplements transport.Transport and performs request over
+// HTTPS with also encoding requests and responses with Base64 encoding.
+type HttpsBase64Transport struct {
```
Comment (nicolai86) [inline transport/https/https.go]: see above comment regarding naming

File: transport/https/https.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+// Do performs the request to the HBCI server. If successful, it returns a
+// populated transport.Response with the HTTP Response Body as Body and the
+// request as Request.
+//
+// Before sending the request it will be encoded with Base64 encoding.
+// When receiving the response with a status code 200 it will decode the response
+// with Base64 encoding. A non 200 status code will be returned as is, without
+// decoding it from Base64.
+func (h *HttpsBase64Transport) Do(request *transport.Request) (*transport.Response, error) {
```
Comment (nicolai86) [inline transport/https/https.go]: see above comment about naming

File: transport/middleware/encoding.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+	"encoding/base64"
+	"io"
+	"io/ioutil"
+
+	"github.com/mitch000001/go-hbci/transport"
+
+	"golang.org/x/text/encoding"
+)
+
+func UTF8Encoding(encoding encoding.Encoding) transport.Middleware {
```
Comment (nicolai86) [inline transport/middleware/encoding.go]: ❤️

**Pull request merged**